File: backend/routes/admin/banners.ts
import { Router, Request, Response } from 'express';
import db from '../../config/database';
import { deleteFromS3 } from '../../config/s3';

const router = Router();

// GET /api/admin/banners
router.get('/', async (_req: Request, res: Response) => {
  try {
    const result = await db.query(
      `SELECT * FROM banners
        ORDER BY sort_order ASC, created_at DESC`,
    );
    return res.json({ success: true, banners: result.rows });
  } catch (err) {
    return res.status(500).json({ success: false, error: (err as Error).message });
  }
});

// POST /api/admin/banners
// Body: { image_url, link_url?, alt_text?, is_active? }
router.post('/', async (req: Request, res: Response) => {
  try {
    const { image_url, link_url, alt_text, is_active } = req.body as {
      image_url?: string;
      link_url?: string | null;
      alt_text?: string | null;
      is_active?: boolean;
    };
    if (!image_url) {
      return res.status(400).json({ success: false, error: 'image_url required' });
    }

    const nextOrder = await db.query(
      'SELECT COALESCE(MAX(sort_order), 0) + 1 AS next FROM banners',
    );

    const result = await db.query(
      `INSERT INTO banners (image_url, link_url, alt_text, is_active, sort_order)
       VALUES ($1, $2, $3, $4, $5)
       RETURNING *`,
      [
        image_url,
        link_url?.trim() ? link_url.trim() : null,
        alt_text?.trim() ? alt_text.trim() : null,
        is_active ?? true,
        nextOrder.rows[0].next,
      ],
    );
    return res.status(201).json({ success: true, banner: result.rows[0] });
  } catch (err) {
    return res.status(500).json({ success: false, error: (err as Error).message });
  }
});

// PUT /api/admin/banners/:id  — partial update; if image_url changes, the
// previous file is removed from S3.
router.put('/:id', async (req: Request, res: Response) => {
  try {
    const { image_url, link_url, alt_text, is_active } = req.body as {
      image_url?: string;
      link_url?: string | null;
      alt_text?: string | null;
      is_active?: boolean;
    };

    const allowed: Record<string, unknown> = {};
    if (image_url !== undefined) allowed['image_url'] = image_url;
    if (link_url  !== undefined) allowed['link_url']  = link_url?.trim() ? link_url.trim() : null;
    if (alt_text  !== undefined) allowed['alt_text']  = alt_text?.trim() ? alt_text.trim() : null;
    if (is_active !== undefined) allowed['is_active'] = is_active;

    if (Object.keys(allowed).length === 0) {
      return res.status(400).json({ success: false, error: 'No fields to update' });
    }

    let previousImage: string | null = null;
    if (image_url !== undefined) {
      const before = await db.query('SELECT image_url FROM banners WHERE id = $1', [req.params.id]);
      previousImage = before.rows[0]?.image_url ?? null;
    }

    const setClauses = Object.keys(allowed).map((k, i) => `${k} = $${i + 1}`);
    const values     = [...Object.values(allowed), req.params.id];
    const result = await db.query(
      `UPDATE banners SET ${setClauses.join(', ')} WHERE id = $${values.length} RETURNING *`,
      values,
    );
    if (!result.rows.length) return res.status(404).json({ success: false, error: 'Not found' });

    if (previousImage && previousImage !== image_url) {
      try { await deleteFromS3(previousImage); }
      catch (s3err) { console.warn('[banners] S3 cleanup failed:', (s3err as Error).message); }
    }

    return res.json({ success: true, banner: result.rows[0] });
  } catch (err) {
    return res.status(500).json({ success: false, error: (err as Error).message });
  }
});

// PATCH /api/admin/banners/reorder
// Body: { items: [{ id, sort_order }, …] }
router.patch('/reorder', async (req: Request, res: Response) => {
  const { items } = req.body as { items?: { id: number; sort_order: number }[] };
  if (!Array.isArray(items)) {
    return res.status(400).json({ success: false, error: 'items array required' });
  }
  const client = await db.pool.connect();
  try {
    await client.query('BEGIN');
    for (const it of items) {
      if (typeof it.id !== 'number' || typeof it.sort_order !== 'number') continue;
      await client.query(
        'UPDATE banners SET sort_order = $1 WHERE id = $2',
        [it.sort_order, it.id],
      );
    }
    await client.query('COMMIT');
    return res.json({ success: true });
  } catch (err) {
    await client.query('ROLLBACK');
    return res.status(500).json({ success: false, error: (err as Error).message });
  } finally {
    client.release();
  }
});

// DELETE /api/admin/banners/:id  — also removes the S3 image (best-effort)
router.delete('/:id', async (req: Request, res: Response) => {
  try {
    const result = await db.query(
      'DELETE FROM banners WHERE id = $1 RETURNING id, image_url',
      [req.params.id],
    );
    if (!result.rows.length) return res.status(404).json({ success: false, error: 'Not found' });

    const imageUrl = result.rows[0].image_url as string | null;
    if (imageUrl) {
      try { await deleteFromS3(imageUrl); }
      catch (s3err) { console.warn('[banners] S3 cleanup failed:', (s3err as Error).message); }
    }
    return res.json({ success: true, message: 'Banner deleted' });
  } catch (err) {
    return res.status(500).json({ success: false, error: (err as Error).message });
  }
});

export default router;

File: backend-golang/internal/handlers/admin/banners.go
package admin

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"

	"github.com/go-chi/chi/v5"

	"inanhxink/backend-golang/internal/config"
	"inanhxink/backend-golang/internal/handlers"
)

// GET /api/admin/banners
func ListBanners(w http.ResponseWriter, r *http.Request) {
	rows, err := config.DB.Query(context.Background(),
		"SELECT * FROM banners ORDER BY sort_order ASC, created_at DESC")
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	banners, err := handlers.CollectRows(rows)
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	handlers.OK(w, map[string]any{"success": true, "banners": banners})
}

// POST /api/admin/banners
func CreateBanner(w http.ResponseWriter, r *http.Request) {
	var body struct {
		ImageURL string  `json:"image_url"`
		LinkURL  *string `json:"link_url"`
		AltText  *string `json:"alt_text"`
		IsActive *bool   `json:"is_active"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil || body.ImageURL == "" {
		handlers.BadRequest(w, "image_url required")
		return
	}

	nextRow := config.DB.QueryRow(context.Background(),
		"SELECT COALESCE(MAX(sort_order), 0) + 1 AS next FROM banners")
	var next int
	nextRow.Scan(&next) //nolint

	isActive := true
	if body.IsActive != nil {
		isActive = *body.IsActive
	}

	cleanLink := nullTrimmed(body.LinkURL)
	cleanAlt := nullTrimmed(body.AltText)

	rows, err := config.DB.Query(context.Background(), `
		INSERT INTO banners (image_url, link_url, alt_text, is_active, sort_order)
		VALUES ($1,$2,$3,$4,$5) RETURNING *`,
		body.ImageURL, cleanLink, cleanAlt, isActive, next)
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	row, _ := handlers.CollectOne(rows)
	handlers.Created(w, map[string]any{"success": true, "banner": row})
}

// PUT /api/admin/banners/:id — partial update; cleans up old S3 image if replaced.
func UpdateBanner(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	var fields map[string]any
	if err := json.NewDecoder(r.Body).Decode(&fields); err != nil {
		handlers.BadRequest(w, "Invalid JSON")
		return
	}

	allowed := map[string]bool{"image_url": true, "link_url": true, "alt_text": true, "is_active": true}
	setClauses := []string{}
	values := []any{}
	i := 1
	for k, v := range fields {
		if !allowed[k] {
			continue
		}
		if k == "link_url" || k == "alt_text" {
			if s, ok := v.(string); ok {
				if strings.TrimSpace(s) == "" {
					v = nil
				} else {
					v = strings.TrimSpace(s)
				}
			}
		}
		setClauses = append(setClauses, fmt.Sprintf("%s = $%d", k, i))
		values = append(values, v)
		i++
	}
	if len(setClauses) == 0 {
		handlers.BadRequest(w, "No fields to update")
		return
	}

	var previousImage string
	if _, replacing := fields["image_url"]; replacing {
		config.DB.QueryRow(context.Background(), //nolint
			"SELECT image_url FROM banners WHERE id = $1", id).Scan(&previousImage)
	}

	values = append(values, id)
	rows, err := config.DB.Query(context.Background(),
		fmt.Sprintf("UPDATE banners SET %s WHERE id = $%d RETURNING *",
			joinClauses(setClauses), len(values)),
		values...)
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	row, err := handlers.CollectOne(rows)
	if err != nil || row == nil {
		handlers.NotFound(w)
		return
	}

	if newImg, _ := fields["image_url"].(string); previousImage != "" && previousImage != newImg {
		config.DeleteFromS3(previousImage) //nolint
	}

	handlers.OK(w, map[string]any{"success": true, "banner": row})
}

// PATCH /api/admin/banners/reorder
func ReorderBanners(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Items []struct {
			ID        int `json:"id"`
			SortOrder int `json:"sort_order"`
		} `json:"items"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		handlers.BadRequest(w, "items array required")
		return
	}

	tx, err := config.DB.Begin(context.Background())
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	defer tx.Rollback(context.Background()) //nolint
	for _, it := range body.Items {
		tx.Exec(context.Background(), //nolint
			"UPDATE banners SET sort_order = $1 WHERE id = $2", it.SortOrder, it.ID)
	}
	if err := tx.Commit(context.Background()); err != nil {
		handlers.InternalError(w, err)
		return
	}
	handlers.OK(w, map[string]any{"success": true})
}

// DELETE /api/admin/banners/:id — also removes S3 image (best-effort)
func DeleteBanner(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	rows, err := config.DB.Query(context.Background(),
		"DELETE FROM banners WHERE id = $1 RETURNING id, image_url", id)
	if err != nil {
		handlers.InternalError(w, err)
		return
	}
	row, err := handlers.CollectOne(rows)
	if err != nil || row == nil {
		handlers.NotFound(w)
		return
	}
	if imgURL, ok := row["image_url"].(string); ok {
		config.DeleteFromS3(imgURL) //nolint
	}
	handlers.OK(w, map[string]any{"success": true, "message": "Banner deleted"})
}

func nullTrimmed(s *string) any {
	if s == nil {
		return nil
	}
	t := strings.TrimSpace(*s)
	if t == "" {
		return nil
	}
	return t
}
